DOC Update web site to pydata-sphinx theme.

cmarmo opened the pull request:
#### What does this implement/fix? Explain your changes.

Dear maintainers,
this draft pull request updates the theme of the website to the [pydata-sphinx](https://github.com/pydata/pydata-sphinx-theme/) theme.
This was suggested [here](https://github.com/scikit-learn/scikit-learn/issues/16951#issuecomment-1604350652). 

Thanks for considering it.

#### Any other comments?

The pull request is a test to check how smooth would be the style conversion, a lot of details need to be fixed but in my opinion the result is already really nice... kudos to the pydata-sphinx theme developers!

I had to update dependencies to add the sphinx theme: my understanding was that I had to run `update_environments_and_lock_files.py` to do that in an automatic way. This ended up in updating the lock files too and I'm not sure it is needed. I was looking for a way to have the CI working ... Is this the way? 

Perhaps @lucyleeow would be interested in having a look and move forward together? :innocent: 

### Functionalities checklist
- [ ] home page carousel
- [ ] what's new labels
- [x] dark theme (see [comment](https://github.com/scikit-learn/scikit-learn/pull/26809#issuecomment-1628571546))
- [x] top bar links (see [comment](https://github.com/scikit-learn/scikit-learn/pull/26809#issuecomment-1628571546))
~~- [ ] old version switch (see [comment](https://github.com/scikit-learn/scikit-learn/pull/26809#issuecomment-1628635403))~~

cmarmo committed:
Update site to pydata-sphinx theme.

lucyleeow commented:
I think `update_environments_and_lock_files.py` is meant to update the lock files but for this PR just updating the doc envs might be enough: `python build_tools/update_environments_and_lock_files.py --select-build doc'

cmarmo committed:
Revert lock files.

cmarmo committed:
Revert lock files.

cmarmo commented:
> `python build_tools/update_environments_and_lock_files.py --select-build doc'

This is the way...

Thanks!

betatim commented:
I think it is a good idea to investigate this. Below a list of things that came to my mind while thinking about this change.

It feels like every so often we have to make changes to the CSS or HTML of the current theme. In those PRs there is often a feeling of "the blind leading the blind". We don't have people who are experts on the topic. This means switching to a theme maintained by people who are (slightly more) expert at it, is a good way to not spend time on something that no one really knows about.

The scikit-learn page has an "iconic look". Hate it or love it, but it is memorable. How much of that can we maintain? It also has a few more things that most documentation pages don't have, or where a given page only has one of the things. This makes me wonder how much work it is to customise the pydata theme, if it is possible and if we don't end up back in the situation where you need to be quite experienced with HTML/CSS in order to fix things. Will it lead to the URLs of pages changing? What do we do about the docs of older versions?

A few years ago when the pydata theme started it felt a bit "unstable". With this I mean when you updated the version things would start working and stop working as well. This was a bit annoying. However in the last couple of years it seems like it has all stabilised a bit more.

cmarmo committed:
[doc build] Update navigation and style of the home page.

cmarmo commented:
> Edit: I realise what you mean. I think the lock files may have updated just because it's been a while since they were last run and there have been new versions of dependencies since

@lucyleeow indeed, but as this is not relevant for this pr I have reverted where possible. However note that updating some dependencies may have led to the failing tests [here](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=56906&view=results).

adrinjalali commented:
> > What do we do about the docs of older versions?
> 
> If I understand correctly older versions will keep their look as they are linked somewhere from the stable one.... but I may be wrong ... I cannot find a lot of explicit documentation about the web infrastructure...

We don't push to the older versions, so they'll stay as they are, which is fine, however, we should make sure switching between versions still works. This new theme also has its own way to handle different versions, I'm not sure if we wanna use it or if we want to have our own version changer thingy there.

cmarmo commented:
@adrinjalali , thanks for your comments

>  It's become a really nice theme and a familiar one to users who use other packages from the stack, which makes it easy for them to navigate the website.

I agree with you, in particular I find the style of the sections less packed and more readable.

> The homepage as is right now in this PR to me is similar enough to what we have, we just need to fix the links from the top bar, and some formatting issues.

I'm working on that, please also note that font-awsome are included in the theme so I have removed the explicit link to github and added the icon in the navigation bar, together with some other social... some other may have their place there, but this is a detail for now.

> We also need to make sure we're dark theme friendly since fabulously this theme respects the browser's setting. Right now it looks like this:

I have added your remarks in the description as this will help me in organizing my time.
Everyone, feel free to add more items to the list there. Thanks!

betatim commented:
> > It also has a few more things that most documentation pages don't have, or where a given page only has one of the things.
> 
> Do you mind giving me an example, so I can address your concerns?

I was thinking of "has quite a custom landing page", "has example gallery, examples are linked to from other parts of the docs with an image preview", "has user guide, examples and API docs", "custom(?) navigation bar/header thing". None of them individually are that complicated/novel, but I don't know many other docs pages that use all of them. So it is more of a vague list of "things to think about and look at" than a list of concrete shortcomings of the theme.

cmarmo committed:
Update navigation bar. Clean styles.

cmarmo committed:
[cd build] Emulate top bar.

cmarmo committed:
[doc build] Fix responsive primary sidebar.

cmarmo commented:
Dear all, the homepage is finalized (the testimonial carousel still need to be fixed, it is just matter of time).
I'll be grateful if you would be able to take some time to browse the site and check if there are broken links or big issues.
Thanks!

GaelVaroquaux commented:
We recently built the website of https://skrub-data.org from the pydata sphinx theme but with a layout somewhat close to that of scikit-learn.

There were some tricky bits on making sure that it was readable both in light and dark background.

The files in skrub can be helpful:
https://github.com/skrub-data/skrub/tree/main/doc

In particular, the files:
* [conf.py](https://github.com/skrub-data/skrub/blob/main/doc/conf.py)
* [_templates/index.html](https://github.com/skrub-data/skrub/blob/main/doc/_templates/index.html)
* Some entries in [_static/css/custom.css](https://github.com/skrub-data/skrub/blob/main/doc/_static/css/custom.css)

Also, one thing to keep in mind is that in scikit-learn we have a ton of custom CSS. Some of it is needed and must stay (eg for the landing page, or for the "details"' folding summary. Some of it will no longer be needed. Making the calls between what should stay and what should go will require iterations

GaelVaroquaux commented:
In my opinion, the current iteration of the landing page does not render well in black
![image](https://github.com/scikit-learn/scikit-learn/assets/208217/a7fcc90e-e30d-4b80-9976-fd9e17517ada)

cmarmo commented:
> In my opinion, the current iteration of the landing page does not render well in black

I should have clarified : the dark theme is not implemented yet (see checklist in the description).
Thanks @GaelVaroquaux for pointing me to the skrub website, I will use the same rendering, otherwise I would have guessed without a clear direction.

I have simplified the theme removing some javascripts and old css which were adding noise in the new scheme.
I'm probably going to remove more... I have heavily modified the structure of the navigation to emulate the current homepage: so for now it really would be helpful if people could check the internal links and navigation here and there, to confirm that nothing is broken.
@lucyleeow , perhaps you can have a look to the example gallery if you have some time? Thanks!

adrinjalali commented:
From what I tested, the links work and the homepage looks quite alright. Links from examples to API docs and to user guide all seem to be fine from what I can tell.

glemaitre commented:
First kudos to @cmarmo on this change. I also like this theme (we used it in `imbalanced-learn`) and it becomes quite a signature theme of the PyData ecosystem.

Making the migration and having all rendering rights will take some time since we don't have testing, and the only way to test is to navigate :). To me, it would make sense to create a new branch and target this PR to this branch. I think it would be beneficial for two reasons:

- we don't want to have a half-working website even on `/dev`
- it will allow targeting the branch with small PRs fixing specific changes that are easy to review instead of having a huge PR

When we are happy with the results then we just merge this branch into main and voila. We should have very few conflicts since we usually don't modify anything from the website HTML/CSS when doing other DOC/code PRs.

betatim commented:
I like the suggestion of targetting a branch. It will let people make small PRs for things they spot instead of one person having to do everything.

How do we setup that branch?

Can we have a rendered version of that branch somewhere that is easy to look at?

GaelVaroquaux commented:
> I like the suggestion of targetting a branch. It will let people make small PRs for things they spot instead of one person having to do everything.

I like that. I'll be able to do PRs :)

glemaitre commented:
> How do we setup that branch?

We only need to create a branch and push it into the repo.

> Can we have a rendered version of that branch somewhere that is easy to look at?

When we will create a PR for the branch, we will have the rendering. We could also modify temporarily the script that pushes the doc into the doc repo to push specifically this branch whenever we merge a PR as well. We could then get the visualization at `https://scikit-learn.org/new_website_branch`

glemaitre commented:
I created `new_web_theme` branch

fuglede committed:
DOC fix description of tolerance in OMP (#26823)

jjerphan committed:
MAINT Remove soon-to-be deprecated scipy.sparse functions (#26751)

cmarmo commented:
> does this setup work for you?

To be honest, I don´t see how adding a new branch change things:
- the rendering needs to be triggered modifying the CI, while here there is no need of changing it
- I'm not expecting this PR to be merged soon
- you all have the rights to push to this PR, anyway. You may also offer PRs to my branch, if you want.
- this wasn't necessary last time the site has been redesigned https://github.com/scikit-learn/scikit-learn/pull/14849
- Occam's razor suggest to minimize the opportunities of messing up with git branching ... ("Entities must not be multiplied beyond necessity")

That said, I guess it's not up to me to decide.

> What do you want to do next?

- 1. Fixi the carousel on the home page
- 2. Clean the file `theme.css` to clarify which issues listed by Guillaume are due to old theme relics, which ones are related to the pydata-sphinx theme
- 3. Define an acceptable dark theme

Reopen this PR versus the new branch, eventually.

glemaitre commented:
> Reopen this PR versus the new branch, eventually.

No need to reopen. You can edit (top-right corner button), and you can actually change the branch that is targeted.

cmarmo commented:
> > Reopen this PR versus the new branch, eventually.
> 
> No need to reopen. You can edit (top-right corner button), and you can actually change the branch that is targeted.

This was my idea of "reopen" ... sorry if it wasn't clear.

cmarmo committed:
Clean theme.css.

cmarmo commented:
I'm marking it as ready, so you can merge at your convenience.

GaelVaroquaux commented:
@cmarmo : do you want to merge to the new_web_theme branch and then we (here an undefined pronoun, as I do not know who will do it) iterate on the PR?

cmarmo commented:
> do you want to merge to the new_web_theme branch and then we (here an undefined pronoun, as I do not know who will do it) iterate on the PR?

I'm going to be explicit... :) .... I'd rather continue to iterate on my branch.
But I'm under the impression that you all want to merge and work directly on the scikit-learn repo ... Feel free to do what you think is the most effective for the task. Thanks.

betatim commented:
A pro of using a branch in the repo here would be that each PR to it would be smaller and easier to review. And once the branch is ready we can merge it without having to look at the huge diff again. Just a thought

cmarmo committed:
[doc build] Cleaning. Implement dark theme.

cmarmo commented:
Dear all, the dark theme is now acceptable.
The main issue with it are the logos because they were not created to be theme-switch proof.
As far as the home page is concerned, I would suggest to remove them from the footer and only add a link to the related page about funding: I'm not planning to reformat and recreate all the logos for both versions (even if this is indeed the right thing to do).

@glemaitre, a number of issues you noticed in [your comment](https://github.com/scikit-learn/scikit-learn/pull/26809#issuecomment-1632452799) are gone, thanks to the fact that I have cleaned up the theme.
About the homepage: "More testimonials" is on the right now, the github icon is on the header (on the sidebar for small screen). To be honest I like it there, that's why I have removed the button.

@betatim,

> A pro of using a branch in the repo here would be that each PR to it would be smaller and easier to review. And once the branch is ready we can merge it without having to look at the huge diff again. Just a thought

Rewriting a website means having a decent homepage and a working general structure, the simpler the better: in my experience this is not something you achieve with small modifications, with the risk of fixing something somewhere and bugging something else elsewhere. One have to be brave at first, and take the leap when the homepage is cleaner enough and there are no more major issues with the structure and the theme. I'm just asking you all to be patient and let me jump. Then you will be free to implement all the small changes you want.

Also, I have checked how the version switcher works: this is when having a branch connected to the sklearn infrastructure will be really useful. I will not take care of that (going to edit the description too) and let those who have the power check the issue.

glemaitre commented:
> @glemaitre, a number of issues you noticed in https://github.com/scikit-learn/scikit-learn/pull/26809#issuecomment-1632452799 are gone, thanks to the fact that I have cleaned up the theme.

Thanks, I will have a new look.

> That said, I guess it's not up to me to decide.

> those who have the power

To be really explicit, I am really offended by these comments that do not bring anything to the discussion.

adrinjalali commented:
This looks great. One thing I notice, is when you hover the mouse over an example in the examples section on the API page, one cannot read the text anymore:

![image](https://github.com/scikit-learn/scikit-learn/assets/1663864/56575c57-2b68-4867-b700-f881629f4d2b)


(also don't love passive aggressive comments)

cmarmo commented:
> To be really explicit, I am really offended by these comments that do not bring anything to the discussion.

@glemaitre , sorry if this blessed you. I really mean lterally. I am unable to check some options because I don't have rights  to access to the infrastructure...

lucyleeow commented:
I think this is stalled and can be taken over if there is still interest.

adrinjalali commented:
@lucyleeow any chance you'd like to take over? 😁

lucyleeow commented:
I'm not very good with (and don't love) html/css...
Maybe someone else if there is still interest in this?

Charlie-XIAO commented:
@lucyleeow I can try to continue this PR if maintainers still want to switch to this new theme  (though I can't guarantee I'm good enough at HTML/CSS).

P.S. I have tried a little bit in my fork https://github.com/Charlie-XIAO/scikit-learn/tree/doc-pydata though it is far from finalized.

lucyleeow commented:
@Charlie-XIAO I think we do want to switch. The only thing I would check is that there are people who have time to review, it may be a big PR.

betatim commented:
In addition to what Lucy wrote: given how big a change this is in terms of how many things need changing and checking, etc I think it would be worth spending a bit of time trying to work out if there is a strategy that allows us to make smaller PRs that can be reviewed and merged independently.

This means coming up with such a strategy would be my first step here. I expect it to be quite hard to come up with the strategy, so while it doesn't involve code it requires some serious amount of skill/experience.

Charlie-XIAO commented:
@betatim Can't agree more, and that's also what I've been concerned about in the past few weeks. It is especially hard to understand what each part of change in a huge PR is for (and because of this I gave up building on top of what @cmarmo has achieved and started from scratch instead).

As for the overall strategy, I would prefer making separate PRs in the main repo, but targeting the `new_web_theme` branch as the current PR is doing. In case maintainers want to know, as follows are the reasons why I do not prefer **one huge PR** (like this one) and **making PRs to my fork** (https://github.com/scikit-learn/scikit-learn/pull/26809#issuecomment-1634331540).

- **One huge PR.** I have been trying in my own fork during waiting for maintainers' response in the past few weeks, so the branch I have been working in is essentially a huge PR. I did try to make things clearer by: (1) breaking down into smaller commits and provide details commit message, and (2) writing comments where I think may be confusing in the source code. If you want, you can look at what I've been doing [here](https://github.com/Charlie-XIAO/scikit-learn/tree/doc-pydata). However, even with these efforts I think the PR would be hard to review as a whole. A previous commit may have flaws and resolved in a later commit, which make changes hard to track, etc.
- **Making PRs to my fork.** Well I first thought this would be nice, but what I'm concerned about is whether CI workflow (in particular doc build) can be triggered in my fork. If not, we will not be able to preview the artifacts. Also, *if* there were other contributors, this may introduce some "overhead" guiding them how to contribute to my fork.

As I've said, I have had a lot of changes ready, and I will definitely break them down so that the review process can be easier. My current plan (draft) is as follows:

- **Switch the theme.** This would be the very first PR, mainly reorganizing `conf.py` to adapt to the new theme, accompanied with other reorganization of files, etc. Also this would involve some setup for following PRs. For instance, I used the `sphinxcontrib.sass` extension so that we can use `sass` for cleaner style sheets, then convert to `css` that can be directly used by `sphinx`, and this setup may need to be done in this PR. This will be a blocking PR, so I would like it to get merged into the `new_web_theme` branch as soon as possible (otherwise the following PRs cannot be started). Correspondingly, I will try to keep it as small as possible (though inevitably it would be at least hundreds of lines of change).

- **Tune the general styles.** This would involve updating `big_toc_css.rst`, fixing the newly added dropdown for the new theme, etc. In other words, these are the general stuff that will not block further PRs but apply to many pages.

- **Version switcher.** The version switcher will need a JSON file (which is essentially a list of version names, version numbers, and urls). The JSON file needs to be at a stable, persistent, fully-resolved URL (i.e., not specified as a path relative to the sphinx root of the current doc build). I tested with [this URL](https://charlie-xiao.github.io/assets/json/sklearn-versions-test.json) hosted on my personal website, but it needs to be finally under `scikit-learn.org`, triggered by something like `build_tools/circle/build_doc.sh`.

- **Tune the landing page.** This would mainly involve modifying `templates/index.html` and add a style sheet for it. The goal is to retain its "iconic look", as you have mentioned. I have almost already accomplished this in both light and dark mode.

  <details>
  <summary>The tuned landing page</summary>
  <p></p>
  <img src="https://github.com/scikit-learn/scikit-learn/assets/108576690/9bfd5f47-ac08-4fe0-85f3-f2010f43c2d3" width="48%" />
  <img src="https://github.com/scikit-learn/scikit-learn/assets/108576690/048a6e90-482b-4aaf-86d4-5498dff9e613" width="48%" />
  </details>

- **More.** Currently I'm moving forward in the order of sections in the header navbar of the current website. If it is a single page (e.g. `install`), then just one PR is sufficient. If it is a collection of pages (e.g. `user_guide`), then maybe we can use an issue to track the changes of its subpages if necessary. Taking `user_guide` as an example, there could be a (meta-)issue tracking the changes of its subpages. For instance: (1) there are many places where we indent the list items, causing them to be enclose in a blockquote (this is not visible in the current theme but `pydata-sphinx-theme` has a background for blockquotes), and we need to fix these; (2) with my current changes we do not need to bold dropdown titles; (3) the citation labels are too messy and it may be nice to change all citations into the format that `pydata-sphinx-theme` can beautify; (4) and many more other details. Then one PR can be made for one or more subpages resolving these details.

Charlie-XIAO commented:
@lucyleeow @betatim Do you recommend me to start opening the first PR now or I still need to wait for more voices?

lucyleeow commented:
That sounds like a reasonable plan to me, but lets hear more voices.

Can I recommend that we move this above plan to a new issue as it would be more suitable and for visibility?

Charlie-XIAO commented:
Sure, I will open an issue for that.
